### ISSUE TITLE
Fails for IE9/10

### ISSUE THREAD

**Issue by simonihmig** [kybishop/ember-popper#42]
See:
* https://travis-ci.org/kaliber5/ember-bootstrap/jobs/275222245
* ![image](https://user-images.githubusercontent.com/1325249/30404265-330493ea-98e6-11e7-9167-e229bcfdd7d0.png)

Looks as if the use of ES6 classes extending from `Ember.Component` is not working here. Not sure why, maybe even some general issue outside the scope of ember-popper?

**Comment by kybishop** [comment]
@pzuraq you're our resident es6 expert. Any thoughts on this/ie9-10 support?

@simonihmig how crucial is ie9/10 support? Seems like their combined market share is down to 1.7% and falling.

**Comment by pzuraq** [comment]
Yeah this seems really weird, looks like Ember is throwing here. I'll need to dive in to see what's up, may have time over the weekend.

If worse comes to worst I'm ok with switching back to .extend, but IE9/10 usage have slipped a lot lately. LI officially only supports IE11, and it could be argued that it's getting closer to becoming a security concern, eol is coming up for both of them I believe

**Comment by simonihmig** [comment]
I did a quick check yesterday, simply extending a component (using Ember object model, and working in IE9+) using an ES6 class, and this failed in IE9 with the same error. So this is for sure a general issue, not specific to ember-popper.

> how crucial is ie9/10 support?

Well, my personal POV is simple: we don't support anything below IE11 for any of our *own* projects (**apps**!). However the situation for an **addon** is somewhat different, as that is something our users have to decide at the end. Sure we could drop support, but maybe some users still have to support older browsers for whatever reason, so those would loose the ability to use that addon (or unable to upgrade).

In my case, I am in the process of refactoring some code to use ember-popper for an addon (ember-bootstrap). So as much as I would *love* to drop support *personally*, I cannot assume this is ok for all users of that addon. So currently I am still trying to support IE9+ as long as possible (or at least as long Ember and Bootstrap themselves do support IE9). In fact I had to do some extra work (e.g. https://github.com/kaliber5/ember-cli-classlist-polyfill) to accomplish that, so I am a bit reluctant to drop support just because of this weird issue we are talking about now...

**Comment by simonihmig** [comment]
> eol is coming up for both of them I believe

They have reached that already afaik! 😬

**Comment by rwjblue** [comment]
Spoke with @simonihmig about this in slack.  The issue here is that `__proto__` is not functional in IE9 and IE10, and babel uses `__proto__` to handle static class methods/properties. To work around the issue you need something like `babel-plugin-transform-proto-to-assign` (or drop IE9 and IE10 support 😈 ).

**Comment by rwjblue** [comment]
Spoke with @simonihmig about this in slack.  The issue here is that `__proto__` is not functional in IE9 and IE10, and babel uses `__proto__` to handle static class methods/properties. To work around the issue you need something like `babel-plugin-transform-proto-to-assign` (or drop IE9 and IE10 support 😈 ).

**Comment by simonihmig** [comment]
> To work around the issue you need something like babel-plugin-transform-proto-to-assign (or drop IE9 and IE10 support 😈 )

Or drop ES6 class syntax here 😬

@rwjblue when we add the plugin [here](https://github.com/kybishop/ember-popper/blob/master/index.js#L44-L60), this will be restricted to transpilation of the addon only, and not fiddle with the consuming app, right?

**Comment by rwjblue** [comment]
> Or drop ES6 class syntax here 😬

Hehe, indeed. Though it seems pretty odd to put too much effort into supporting IE9/IE10.  They are unsupported _and_ have super low market share...

> @rwjblue when we add the plugin here, this will be restricted to transpilation of the addon only, and not fiddle with the consuming app, right?

Confirm.

**Comment by pzuraq** [comment]
I'm torn on this a few ways. On the one hand I don't think the babel plugin should be added here, if anything is should be added to `ember-decorators` since this is more of an issue with the babel setup it adds. It should also be optional, since not all users want/need to support IE 9/10.

On the other hand, I hate supporting older IE versions - it just extends their lifespan and adds tech debt, and we're _so_ close to all evergreen forever. 

@simonihmig do you have a solid requirement to support IE 9/10? If so I think it's worth it to add this as an option upstream in `ember-decorators`, but if not I'd like to hold off personally

**Comment by kybishop** [comment]
+1. Seems like, if anything, this is an issue upstream. Going to close this for now, but feel free to discuss more here before we move it up the chain.

**Comment by simonihmig** [comment]
> if anything is should be added to ember-decorators since this is more of an issue with the babel setup it adds.

I don't see how this is related to `ember-decorators` at all!? @rwjblue created [this demo](http://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&code_lz=MYGwhgzhAEBiD29oG8BQ1oQC5iwS2GgHMBTLaPCAYXgFsAHeAOxKawAoBKFaAJzICuvJtCy8BJANzQAvqjmpQkGACEwvaCQAeWVgBMYCJMgXBmEeCBIA6EPCLs1va5RoNmrLJ0lA&debug=false&circleciRepo=&evaluate=true&lineWrap=false&presets=es2015%2Cstage-2%2Cstage-3&prettier=false&targets=&version=6.26.0) in the slack conversation that shows the issue. The transpiled code outputs `undefined` in IE9, and does *not* use any decorators! As he stated, the issue is how Babel transpiles ES6 classes, which make use of `__proto__` that is not supported in IE9+10.

> On the other hand, I hate supporting older IE versions - it just extends their lifespan and adds tech debt, and we're so close to all evergreen forever.

In general I agree 💯  *personally*. But as I stated my motivation for supporting those older IE versions [here](https://github.com/kybishop/ember-popper/issues/42#issuecomment-329995090), I don't think it is *my/our* call to make. I would put *user* needs here above *maintainer* needs, and let our users decide. Dropping support would be a breaking change for my addon, which I don't want to do without a good reason. Fwiw, ember itself still *has* to support IE9, any change regarding that would require a major version bump, even though certainly it would make maintaining it a bunch easier...

One thing worth mentioning here is that popper.js explicitly supports IE10 (and tests this), so for a "wrapper" addon which this one is, it seems strange to me to support less browsers than the original library.

My suggestion would be to add the mentioned babel plugin here. Maybe this can be done just if needed, by checking the EmberCLI targets. Afaik the canisue data does not cover JS support, only browser APIs, but we probably can check if the browserlist itself includes IE9 or 10 explicitly. But at the end I think the additional code added by that plugin is probably just a few lines, so I don't really see the big downsides here, even if done unconditionally. 

It's your call at the end, if you don't agree that's fine! But I will then probably have to go for a fork, or integrate popper.js directly in some way...

**Comment by pzuraq** [comment]
My point is we don't have a user with these needs yet, so we shouldn't prematurely optimize. As soon as a user asks for this functionality, I'm happy to add it.

As for adding it to ember-decorators, you're correct there. My logic was that anyone who was using classes would also be using ember-decorators since it's pretty hard to be effective without them. It was also to avoid every single addon author from having to duplicate this code when using classes, since this will be a common use case, plus as I said before it should be opt-in meaning there should be a single standardized config option that can be passed into the app rather than one per-addon. I could also see it being a separate addon, the only downside being if an addon author forgets it, it won't be compatible - simple PR though.

**Comment by srvance** [comment]
> My point is we don't have a user with these needs yet, so we shouldn't prematurely optimize. As soon as a user asks for this functionality, I'm happy to add it.

I would support @simonihmig on this and contend that ember-bootstrap is bringing many users to ember-popper with these needs. We don't need to wait for them to scream, and in fact, as an established addon, it would be irresponsible to make them scream before addressing this.

I like the idea of babel not causing this problem in the first place, but it seems like something that should be worked around until such time we or someone else addresses the deeper issue and we can remove the workaround.

**Comment by pzuraq** [comment]
Totally, then you're the user asking for the functionality 🙂 We'll work out a solution upstream somewhere, be it in `ember-decorators` or a different addon, and integrate it here

**Comment by simonihmig** [comment]
I am not sure how Babel transpilation should work when pushing this upstream. As confirmed [here](https://github.com/kybishop/ember-popper/issues/42#issuecomment-330263726), we can make this plugin work just on the addon's tree, which seams reasonable to me. Making it do its work on the whole consuming app should be prevented I *think*, as it may not use ES6 classes at all and thus does not need it. So if we just want to make its transpilation works on the addon's files only, we have to add it here, not somewhere upstream, or do I miss something?

@rwjblue as you were already involved here and maintain ember-decorators itself, may I ask you on your optioning about this?

**Comment by srvance** [comment]
> Totally, then you're the user asking for the functionality 🙂

👍 And on behalf of many others.

**Comment by pzuraq** [comment]
We discussed this in Slack, but the issue is that users will want this to apply to all addons and the app, not just this one. `ember-decorators` adds the transforms _just_ to its parent addon/app, which is how the upstream solution would work.

**Comment by simonihmig** [comment]
May I ask how we can move this forward? As we discussed in Slack, it may be reasonable to add this plugin to `ember-cli-babel`, but I don't know how fast this will happen, and I am not sure I will be able to contribute this in the near future. 

So as we discussed this quite a bit and invested some time, and this really is something that probably can be easily fixed and is still a *blocking* issue for me, I would love to move this forward in a pragmatic way. So would you guys accept a PR that adds this plugin to this addon for the time being, until a long term solution somewhere upstream is available?

**Comment by kybishop** [comment]
@simonihmig Sorry for keeping you blocked! I'd love to have a solution for this that makes everyone happy. My gut says yes but I'd like to defer to @pzuraq since he understands the implications better than me.

**Comment by pzuraq** [comment]
yeah it's reasonable to add it here if you're blocked on it. Feel free to submit the PR, we can merge and release as soon as it's ready.

**Comment by simonihmig** [comment]
Thanks for the super quick response! Will do this soon, and make sure it runs fine in those godd****d IE! ;)

**Comment by simonihmig** [comment]
Oh man, this drives me nuts... I tried this yesterday, but to no avail. Adding the plugin worked so far, and it did fix Ember throwing that type assertion. But then I see this error thrown somewhere within Glimmer code!? (*that's IE10 in this case*)

![image](https://user-images.githubusercontent.com/1325249/30734145-4752b3aa-9f7a-11e7-9876-a2f3b97e9480.png)

I don't know... I quickly refactored the components to plain Ember object model, and that worked flawlessly, all tests passing... currently I don't see any better way to move this forward as a short term solution, so would you accept such a PR? 

If you feel that's not the direction you want to go, and support for those rotten IEs is not a high priority from your POV, then I might still go the fork road, and create something like `ember-popper-combat` for my specific compatibility requirements, but for sure I am not very eager to do that because of the  additional maintenance burden...

**Comment by pzuraq** [comment]
Can you push the branch with the plugin so I can take a look?

**Comment by simonihmig** [comment]
@pzuraq sure: https://github.com/simonihmig/ember-popper/tree/fix-ie

I also tried adding it as a `postTransformPlugin`, but that made no difference...

**Comment by pzuraq** [comment]
Ok, will take a look when I get a chance. Does this also fail for IE10? It looks like Bootstrap has dropped support for IE9, so I think there's even less reason to go back that far.

**Comment by simonihmig** [comment]
> Does this also fail for IE10

Yes, the screenshot above is IE10!

**Comment by pzuraq** [comment]
👍 Cool, I'll dive in once I get the chance, probably Monday (all booked today).

**Comment by simonihmig** [comment]
Given that the days of IE9/10 support in Ember itself seem to be counted (https://github.com/emberjs/rfcs/pull/252) - for good reasons - it maybe makes sense to rethink our requirements here. I would try to get some feedback from ember-bootstrap users on their requirements, and see what we can learn from that... so maybe we can put this issue on hold for now!

**Comment by simonihmig** [comment]
> I would try to get some feedback from ember-bootstrap users on their requirements

As it turned out, nobody seemed to care about IE9/10 anymore. Which is nice, as we can bury this issue here finally!

However I really feel sorry for having all of you bugged so long about this. Quite needlessly as it turned out... Owe you a 🍺! 😋

**Comment by kybishop** [comment]
@simonihmig no worries, glad it all worked out!
